### PR TITLE
Update trigger threshold guess estimation to mean of trigger channel

### LINKED ISSUE #211: Improve trigger threshold computation
## Detailed Description

While working on #206 @sangfrois and I realised that the current approach to compute the threshold fails for multi-runs recordings. Digging into the matter a bit deeper, I realised that this happens if, in absence of noise:
- The average is different from zero and there are too many peak points, compared to the baseline points.
- The average is different from zero and is considerable compared to the level the peaks get to.

As an easy fix, we could set the threshold to be one standard deviation from the average, instead of two standard deviations from the average.
Testing it a bit more with noisy signal, I realised that one standard deviation from average wouldn't be enough to proper threshold points if:
- the noise is considerable compared to the level the peaks get to
- the noise is not much, but there are not enough peak points
- the noise is not much, but the average is not around zero.

A less easy fix in this case would be reconsidering the options in #173 and implement a different approach.

## Possible Implementation

1. We could simply retune the threshold as one standard deviation from the average of the trigger channel, OR
2. We could implement a different approach, for instance the one based on distributions, OR
3. We could do both, keep the standard deviation as "default" approach, and implement a different approach upon user request. 

In case we decide to go for option 1, please relabel this issue as bug and the PR as BugFIX

### PR DESCRIPTION
**Pull request by 62442katieb**
Closes #211 

## Proposed Changes

  - Trigger threshold is now defined as the mean value of the trigger channel across time
  - Logic: if trigger channel is binary (or roughly binary), the "on" state is always greater than the mean and the "off" state is always less than the mean

## Change Type

- [x] `bugfix` (+0.0.1)
- [ ] `minor` (+0.1.0)
- [ ] `major`  (+1.0.0)
- [ ] `refactoring` (no version update)
- [ ] `test` (no version update)
- [ ] `infrastructure` (no version update)
- [ ] `documentation` (no version update)
- [ ] `other`

## Checklist before review

- [x] I added everything I wanted to add to this PR.
- [x] \[Code or tests only\] I wrote/updated the necessary docstrings.
- [ ] \[Code or tests only\] I ran and passed tests locally.
- [ ] \[Documentation only\] I built the docs locally.
- [x] My contribution is harmonious with the rest of the code: I'm not introducing repetitions.
- [x] My code respects the adopted style, especially linting conventions.
- [x] The title of this PR is explanatory on its own, enough to be understood as part of a changelog.
- [x] I added or indicated the right labels.

- [ ] I added information regarding the timeline of completion for this PR.

- [ ] Please, comment on my PR while it's a draft and give me feedback on the development!

### CODE REVIEW

File: phys2bids/physio_obj.py
Comment (vinferrer) [inline phys2bids/physio_obj.py:473]: If the threshold is gonna be the mean of the trigger, this line is not used at all then. Could you delete it?

**Review by vinferrer — CHANGES_REQUESTED** [review]
Also you should look into the docs and update them, For example: [Here in line 205](https://github.com/62442katieb/phys2bids/blob/bug/trigger-thresh/docs/howto.rst ). There is also a linting error, seems like you left white spaces at the end of some lines. After those changes, it LGTM for approval

File: phys2bids/physio_obj.py
Comment (62442katieb) [reply phys2bids/physio_obj.py]: Oops! Yes, will do. that was a copy/paste error.

**Comment by 62442katieb** [comment]
> Also you should look into the docs and update them, For example: [Here in line 205](https://github.com/62442katieb/phys2bids/blob/bug/trigger-thresh/docs/howto.rst). 
Good catch, fixed! 👍 

>There is also a linting error, seems like you left white spaces at the end of some lines. After those changes, it LGTM for approval
Linting errors also fixed 👍

**Comment by vinferrer** [comment]
LGTM!! @smoia do you think this is a Bugfix? in that case we don't need a second review

**Review by vinferrer — APPROVED** [review]
LGTM

**Review by smoia — COMMENTED** [review]
I'd say it 's LeGiTiMate, but I'm probably over-worrying about very long multi-run recordings where you have a lot of time between runs and the mean might be too sensitive to noise. 

May I ask @sangfrois or @RayStick or anyone else that has a good, long multi-run session with a noisy trigger channel at disposal to run this PR and communicate the results?

**Comment by RayStick** [comment]
I'm happy to do some testing (but I can only do that Tues 26th, in case there is a specific timeline needed)

**Comment by smoia** [comment]
No timeline, go for it!

**Comment by RayStick** [comment]
I'm having trouble with my python/phys2bids set-up at the moment - sorry! I can check this again on Wednesday to see if i can troubleshoot it but the data I was going to test on OSF**: labchart/chicago/four_scans_samefreq_allchannels.txt** and has a json file.

**Comment by sangfrois** [comment]
Hey @RayStick ! No problem, I got this.
I tested it with a multi-run acknowledge file. It has 8 runs. It spans about 1h30. 


### The trigger signal looks like this 
**Note that Y axis is Voltage**
image
### The trigger actually looks like this
image
image
@smoia the noise seems to be trivial. So I wouldn't be worried to derive a threshold based on the mean that would confound noise and real triggers.

### The given input is
``phys2bids -in path/to/multi-run.acq -ntp 406 406 406 3 410 410 410 2 -tr 1.49 -chtrig 4``
### The output is
```
INFO       Currently running phys2bids version 2.2.3+725.g3b4ec90.dirty
INFO       Input file is C:/Users/sangfrois/Desktop/NeuroMod_tests/multi-run.acq
INFO       File extension is .acq
WARNING    If both acq and txt files exist in the path, acq will be selected.
INFO       Reading the file C:\Users\sangfrois\Desktop\NeuroMod_tests\multi-run.acq
INFO       00. Channel GSR-EDA100C-MRI: 26212257 samples, 5000.0 samples/sec, loaded: True
INFO       01. Channel PPG100C: 26212257 samples, 5000.0 samples/sec, loaded: True
INFO       02. Channel ECG100C: 26212257 samples, 5000.0 samples/sec, loaded: True
INFO       03. Channel TTL: 26212257 samples, 5000.0 samples/sec, loaded: True
INFO       04. Channel Custom, HLT100C - A 6: 26212257 samples, 5000.0 samples/sec, loaded: True
INFO       Trigger channel name is not in our trigger channel name alias list. Please make sure you choose the proper channel.
INFO       Checking that units of measure are BIDS compatible
INFO       Reading infos
INFO
------------------------------------------------
File C:\Users\sangfrois\Desktop\NeuroMod_tests\multi-run.acq contains:
01. GSR-EDA100C-MRI; sampled at 5000.0 Hz
02. PPG100C; sampled at 5000.0 Hz
03. ECG100C; sampled at 5000.0 Hz
04. TTL; sampled at 5000.0 Hz
05. Custom, HLT100C - A 6; sampled at 5000.0 Hz
------------------------------------------------

INFO       saving channel plot to C:\Users\sangfrois\git\phys2bids\code\conversion\multi-run.png
INFO       Counting trigger points
INFO       The trigger is in channel 4
INFO       The number of timepoints according to the std_thr method is 2453. The computed threshold is 0.3033
INFO       Checking number of timepoints
INFO       Found just the right amount of timepoints!
WARNING
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
phys2bids will split the input file according to the given -tr and -ntp arguments
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
INFO       Counting trigger points
INFO       The trigger is in channel 4
INFO       The number of timepoints found with the manual threshold of 0.3033 is 2453
INFO       Checking number of timepoints
WARNING    Found 2047 timepoints more than expected!
Assuming extra timepoints are at the end (try again with a more liberal thr)
INFO
--------------------------------------------------------------
Slicing between 0.0 seconds and 1328.1316 seconds
--------------------------------------------------------------
INFO       Trigger channel name is not in our trigger channel name alias list. Please make sure you choose the proper channel.
INFO       Counting trigger points
INFO       The trigger is in channel 4
INFO       The number of timepoints found with the manual threshold of 0.3033 is 2047
INFO       Checking number of timepoints
WARNING    Found 1641 timepoints more than expected!
Assuming extra timepoints are at the end (try again with a more liberal thr)
INFO
--------------------------------------------------------------
Slicing between 1346.728 seconds and 1969.668 seconds
--------------------------------------------------------------
INFO       Trigger channel name is not in our trigger channel name alias list. Please make sure you choose the proper channel.
INFO       Counting trigger points
INFO       The trigger is in channel 4
INFO       The number of timepoints found with the manual threshold of 0.3033 is 1641
INFO       Checking number of timepoints
WARNING    Found 1235 timepoints more than expected!
Assuming extra timepoints are at the end (try again with a more liberal thr)
INFO
--------------------------------------------------------------
Slicing between 2006.4526 seconds and 2629.3926 seconds
--------------------------------------------------------------
INFO       Trigger channel name is not in our trigger channel name alias list. Please make sure you choose the proper channel.
INFO       Counting trigger points
INFO       The trigger is in channel 4
INFO       The number of timepoints found with the manual threshold of 0.3033 is 1235
INFO       Checking number of timepoints
WARNING    Found 1232 timepoints more than expected!
Assuming extra timepoints are at the end (try again with a more liberal thr)
INFO
--------------------------------------------------------------
Slicing between 2624.3546 seconds and 2646.8246 seconds
--------------------------------------------------------------
INFO       Trigger channel name is not in our trigger channel name alias list. Please make sure you choose the proper channel.
INFO       Counting trigger points
INFO       The trigger is in channel 4
INFO       The number of timepoints found with the manual threshold of 0.3033 is 1232
INFO       Checking number of timepoints
WARNING    Found 822 timepoints more than expected!
Assuming extra timepoints are at the end (try again with a more liberal thr)
INFO
--------------------------------------------------------------
Slicing between 3320.5608 seconds and 3949.4608 seconds
--------------------------------------------------------------
INFO       Trigger channel name is not in our trigger channel name alias list. Please make sure you choose the proper channel.
INFO       Counting trigger points
INFO       The trigger is in channel 4
INFO       The number of timepoints found with the manual threshold of 0.3033 is 822
INFO       Checking number of timepoints
WARNING    Found 412 timepoints more than expected!
Assuming extra timepoints are at the end (try again with a more liberal thr)
INFO
--------------------------------------------------------------
Slicing between 3962.286 seconds and 4591.186 seconds
--------------------------------------------------------------
INFO       Trigger channel name is not in our trigger channel name alias list. Please make sure you choose the proper channel.
INFO       Counting trigger points
INFO       The trigger is in channel 4
INFO       The number of timepoints found with the manual threshold of 0.3033 is 412
INFO       Checking number of timepoints
WARNING    Found 2 timepoints more than expected!
Assuming extra timepoints are at the end (try again with a more liberal thr)
INFO
--------------------------------------------------------------
Slicing between 4606.9372 seconds and 5235.8372 seconds
--------------------------------------------------------------
INFO       Trigger channel name is not in our trigger channel name alias list. Please make sure you choose the proper channel.
INFO       Counting trigger points
INFO       The trigger is in channel 4
INFO       The number of timepoints found with the manual threshold of 0.3033 is 2
INFO       Checking number of timepoints
INFO       Found just the right amount of timepoints!
INFO
--------------------------------------------------------------
Slicing between 5230.9196 seconds and 5242.4486 seconds
--------------------------------------------------------------
INFO       Trigger channel name is not in our trigger channel name alias list. Please make sure you choose the proper channel.
INFO       Trigger channel name is not in our trigger channel name alias list. Please make sure you choose the proper channel.
INFO       Counting trigger points
INFO       The trigger is in channel 4
INFO       The number of timepoints found with the manual threshold of 0.3033 is 406
INFO       Checking number of timepoints
INFO       Found just the right amount of timepoints!
INFO       Trigger channel name is not in our trigger channel name alias list. Please make sure you choose the proper channel.
INFO       Counting trigger points
INFO       The trigger is in channel 4
INFO       The number of timepoints found with the manual threshold of 0.3033 is 406
INFO       Checking number of timepoints
INFO       Found just the right amount of timepoints!
INFO       Trigger channel name is not in our trigger channel name alias list. Please make sure you choose the proper channel.
INFO       Counting trigger points
INFO       The trigger is in channel 4
INFO       The number of timepoints found with the manual threshold of 0.3033 is 406
INFO       Checking number of timepoints
INFO       Found just the right amount of timepoints!
INFO       Trigger channel name is not in our trigger channel name alias list. Please make sure you choose the proper channel.
INFO       Counting trigger points
INFO       The trigger is in channel 4
INFO       The number of timepoints found with the manual threshold of 0.3033 is 3
INFO       Checking number of timepoints
INFO       Found just the right amount of timepoints!
INFO       Trigger channel name is not in our trigger channel name alias list. Please make sure you choose the proper channel.
INFO       Counting trigger points
INFO       The trigger is in channel 4
INFO       The number of timepoints found with the manual threshold of 0.3033 is 410
INFO       Checking number of timepoints
INFO       Found just the right amount of timepoints!
INFO       Trigger channel name is not in our trigger channel name alias list. Please make sure you choose the proper channel.
INFO       Counting trigger points
INFO       The trigger is in channel 4
INFO       The number of timepoints found with the manual threshold of 0.3033 is 410
INFO       Checking number of timepoints
INFO       Found just the right amount of timepoints!
INFO       Trigger channel name is not in our trigger channel name alias list. Please make sure you choose the proper channel.
INFO       Counting trigger points
INFO       The trigger is in channel 4
INFO       The number of timepoints found with the manual threshold of 0.3033 is 410
INFO       Checking number of timepoints
INFO       Found just the right amount of timepoints!
INFO       Trigger channel name is not in our trigger channel name alias list. Please make sure you choose the proper channel.
INFO       Counting trigger points
INFO       The trigger is in channel 4
INFO       The number of timepoints found with the manual threshold of 0.3033 is 2
INFO       Checking number of timepoints
INFO       Found just the right amount of timepoints!
INFO       Plot trigger
INFO       Plot trigger
INFO       Plot trigger
INFO       Plot trigger
INFO       Plot trigger
INFO       Plot trigger
INFO       Plot trigger
INFO       Plot trigger
INFO       Found 8 different scans in input!
INFO       Preparing 8 output files.
INFO       Exporting files for run 1 freq 5000.0
```

### The code behaves properly
This needs to be merged. Changes are parcimonious and well structure. Though, I have a little comment :  

``INFO       The number of timepoints according to the std_thr method is 2453. The computed threshold is 0.3033``

What is std_thr method actually ? Does it stand for standard deviation ? If yes, I imagine is related to the current trigger deteciton method in main version. So, I suggest we change the log. Just hit 👍 if you want me to do it!

**Comment by sangfrois** [comment]
Also, might I point out that the Y axis on the visualization does not have the same shape as in the source file. On the source file, the trigger values range from 0 to 5 (as you can see on previous graphs), but our trigger plot gives us values ranging from 0 to 1.

image

Is this normal ?

**Comment by RayStick** [comment]
@sangfrois thanks!

**Comment by smoia** [comment]
I guess we can merge this PR at this point - @RayStick @sangfrois are we happy with the results?

**Comment by sangfrois** [comment]
@smoia Totally! :)

**Comment by RayStick** [comment]
I'm sorry I haven't done any testing yet :pensive:
If it works for @sangfrois I'd say merge it, but I can still test down the line

**Pull request merged**
